// zigbee-herdsman-converters/devices/eglo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const extend_1 = __importDefault(require("../lib/extend"));
const fromZigbee_1 = __importDefault(require("../converters/fromZigbee"));
const exposes = __importStar(require("../lib/exposes"));
const e = exposes.presets;
const definitions = [
    {
        zigbeeModel: ['EBF_RGB_Zm', 'EBF_RGB_Zm_CLP'],
        model: '900091',
        vendor: 'EGLO',
        description: 'ROVITO-Z ceiling light',
        extend: extend_1.default.light_onoff_brightness_colortemp_color({ colorTempRange: [153, 370] }),
    },
    {
        zigbeeModel: ['ESMLFzm_w6_TW'],
        model: '12242',
        vendor: 'EGLO',
        description: 'ST64 adjustable white filament bulb',
        extend: extend_1.default.light_onoff_brightness_colortemp({ colorTempRange: [153, 454] }),
    },
    {
        zigbeeModel: ['EGLO_ZM_RGB_TW'],
        model: '900024/12253',
        vendor: 'EGLO',
        description: 'RGB light',
        extend: extend_1.default.light_onoff_brightness_colortemp_color({ colorTempRange: [153, 370], supportsHueAndSaturation: true }),
    },
    {
        zigbeeModel: ['EGLO_ZM_TW_CLP'],
        model: '98847',
        vendor: 'EGLO',
        description: 'FUEVA-Z ceiling light IP44',
        extend: extend_1.default.light_onoff_brightness_colortemp({ colorTempRange: [153, 370] }),
    },
    {
        zigbeeModel: ['ERCU_3groups_Zm'],
        model: '99099',
        vendor: 'EGLO',
        description: '3 groups remote controller',
        fromZigbee: [fromZigbee_1.default.command_on, fromZigbee_1.default.awox_colors, fromZigbee_1.default.awox_refresh, fromZigbee_1.default.awox_refreshColored, fromZigbee_1.default.command_off,
            fromZigbee_1.default.command_step, fromZigbee_1.default.command_move, fromZigbee_1.default.command_move_to_level, fromZigbee_1.default.command_move_to_color_temp,
            fromZigbee_1.default.command_stop, fromZigbee_1.default.command_recall, fromZigbee_1.default.command_step_color_temperature],
        toZigbee: [],
        exposes: [e.action(['on', 'off', 'red', 'refresh', 'refresh_colored', 'blue', 'yellow',
                'green', 'brightness_step_up', 'brightness_step_down', 'brightness_move_up', 'brightness_move_down', 'brightness_stop',
                'recall_1', 'color_temperature_step_up', 'color_temperature_step_down'])],
    },
];
exports.default = definitions;
module.exports = definitions;
